### PR TITLE
AES-XTS VFS

### LINKED ISSUE #167: AES-XTS VFS
When I was going through a crypto review as part of internal corporate processes, the reviewers found the crypto in Adiantum a bit too new-fangled. Nothing inherently wrong, just not on a pre-vetted list.

As such, it was suggested that I use AES-XTS and key generation via PBKDF2 instead of Argon2id. My XTS VFS implementation is not highly differentiated from the Adiantum VFS, except perhaps that it only imports `x/crypto`. 

Are you interested in an XTS VFS PR (with `internal/util.AssertErr` added)?

See https://github.com/fido-device-onboard/go-fdo/blob/main/sqlite/xts/xts.go

### PR DESCRIPTION
**Pull request by ncruces**
Fix #167.

Thank you, @ben-krieger!

### CODE REVIEW

File: vfs/xts/api.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"golang.org/x/crypto/xts"
+)
+
+func init() {
+	Register("xts", vfs.Find(""), nil)
+}
+
+// Register registers an encrypting VFS, wrapping a base VFS,
+// and possibly using a custom XTS cipher construction.
+// To use the default AES-XTS construction, set cipher to nil.
```
Comment (ben-krieger) [inline vfs/xts/api.go:47]:
```suggestion
// Register registers an encrypting VFS, wrapping a base VFS,
// and possibly using a custom XTS cipher construction.
// To use the default AES-XTS construction, set cipher to nil.
//
// The default cipher uses AES-128, AES-192, or AES-256 if the
// key/hexkey is 16, 24, or 32 bytes, respectively. If a textkey
// is provided, the default KDF is PBKDF2-HMAC-SHA512 with
// 10,000 iterations, always producing a 32 byte key.
```

File: vfs/xts/xts.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
+package xts
+
+import (
+	"encoding/hex"
+	"io"
+
+	"github.com/ncruces/go-sqlite3"
+	"github.com/ncruces/go-sqlite3/internal/util"
+	"github.com/ncruces/go-sqlite3/vfs"
+	"golang.org/x/crypto/xts"
```
Comment (ben-krieger) [inline vfs/xts/xts.go:10]:
```suggestion
	"golang.org/x/crypto/xts"

	"github.com/ncruces/go-sqlite3"
	"github.com/ncruces/go-sqlite3/internal/util"
	"github.com/ncruces/go-sqlite3/vfs"
```

Do you care about formatting matching `goimports -local github.com/ncruces/go-sqlite3`? If using `golangci-lint`, you'd add this to catch:

```yaml
linters-settings:
  goimports:
    local-prefixes: github.com/ncruces/go-sqlite3
```

File: vfs/xts/xts.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+func (x *xtsFile) Pragma(name string, value string) (string, error) {
+	var key []byte
+	switch name {
+	case "key":
+		key = []byte(value)
+	case "hexkey":
+		key, _ = hex.DecodeString(value)
+	case "textkey":
+		key = x.init.KDF(value)
```
Comment (ben-krieger) [inline vfs/xts/xts.go:78]: Is there any legitimate case for a random key? If not, then possibly error if value is `""`... or at least warn?

File: vfs/xts/xts.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+		cipher = x.init.XTS(key)
+	}
+
+	if cipher == nil {
+		return nil, flags, sqlite3.CANTOPEN
+	}
+	return &xtsFile{File: file, cipher: cipher, init: x.init}, flags, nil
+}
+
+const sectorSize = 512
```
Comment (ben-krieger) [inline vfs/xts/xts.go:61]: Did you want to leave a comment that this isn't an XTS requirement, but doesn't seem to matter for performance?

**Review by ben-krieger — APPROVED** [review]

**Comment by ben-krieger** [comment]
Behavior seems to be the same as before with https://github.com/fido-device-onboard/go-fdo/pull/11, but I can't directly compare, since the pepper and sector size changed.

File: vfs/xts/api.go (hunk 1 above)
Comment (ncruces) [reply vfs/xts/api.go:47]: XTS needs double sized keys, so it's 32 to 64 bytes, actually.

File: vfs/xts/xts.go (hunk 3 above)
Comment (ncruces) [reply vfs/xts/xts.go:78]: Yes. Temporary files use a random key. But probably not when specified as a `textkey`. Let me check the code paths.

File: internal/util/math_test.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package adiantum
+package util
```
[the remark this answers is not in the document]
Comment (ncruces) [reply internal/util/math_test.go:1]: I have a test for the `abs(-1)` corner case, and I didn't export `abs`.

File: vfs/xts/xts.go (hunk 3 above)
Comment (ncruces) [reply vfs/xts/xts.go]: Good catch!

**Comment by ncruces** [comment]
@ben-krieger the fact that I merged this doesn't mean that it's final (even if I tag a release).
I'll mention in the release notes that this VFS is experimental, until you give final approval.

If security review on your end feels that anything should change, and that requires a compatibility break, we do it anyway.
Security review is kinda whole the point of this, and `go-fdo` the primary client.
There's little point in keeping compatibility with something that fails review and goes unused.

File: vfs/xts/xts.go (hunk 2 above)
Comment (ncruces) [reply vfs/xts/xts.go]: Thanks! See: ace01b2.

**Comment by ben-krieger** [comment]
> @ben-krieger the fact that I merged this doesn't mean that it's final (even if I tag a release). I'll mention in the release notes that this VFS is experimental, until you give final approval.
> 
> If security review on your end feels that anything should change, and that requires a compatibility break, we do it anyway. Security review is kinda whole the point of this, and `go-fdo` the primary client. There's little point in keeping compatibility with something that fails review and goes unused.

I'm trying to get an informal review done now. But worst case, there will be another official review once we tag the first release of `go-fdo`, so I'll keep you updated.
